docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+- The OOM killer might kill processes across different pods. Suppose that 10 best-effort pods are running on the Kubelet, and the system runs out of memory. The OOM killer might kill one process from each pod. Short-term TODO: Docker’s philosophy has been not to care about this issue, so for now we will not focus on it. Long-term TODO: Treat a pod as a unit - so that one entire pods is killed, instead of a few processes across different pods.
+- CPU shares allocation issue. Suppose that a node is running 2 pods: a Burstable pod requesting for 50% of CPU (but without a CPU limit), and a Best-Effort pod. Suppose that both pods try to use as much CPU as possible. After the proposal is implemented, the Burstable pod will get 100% of the CPU, and the Best-Effort pod will get around 0% of the CPU. However, a fairer scheme would give the Burstable pod 75% of the CPU and the Best-Effort pod 25% of the CPU (since resources past the Burstable pod’s request are not guaranteed). Long-term TODO: Think of a way to solve this problem.
+- Ensuring that the proposal works in systemd platforms might be tricky, because they don’t allow explicit manipulation of the cgroup hierarchy. Handling distributions of Linux based on systemd is critical, because major Linux distributions like Debian and Ubuntu are moving to systemd. TODO: investigate challenges and ensure that the proposal works on systemd systems.
+
+## Demo and Tests
+
+I will implement a prototype and demo the prototype at a Kubernetes Community meeting. Things I will show in the demo:
+- Launch a couple of top-tier pods on a node. Barrage the node with best-effort pods. The top-tier pods should survive the onslaught of best-effort pods.
+- Fill up a node with best-effort pods. Barrage the node with top-tier pods. All best-effort pods should be evicted.
+- Launch a best-effort pod that tries to use up CPU resources. The best-effort pod, when run in isolation, should get to use the entire CPU. Then add a top-tier pod that tries to use up CPU. The best-effort pod should be throttled, and given a small number of CPU shares.
+
+E2E tests for the system can be flaky, however I will write unit tests for the various components.
+
+## Milestones
+
+1. Add “policy” code that decides what tier a pod is in, and what OOM scores/CPU shares a pod should get
+2. Add a module to set the OOM score of a pod
+3. Implement add pod (to set CPU shares, OOM scores, CPU quotas correctly)
+4. Set the OOM score of infrastructure pods, Kubelet, Kube-proxy, etc, appropriately
+5. Modify pod admission code in scheduler to use request, and to deal with multiple tiers
+6. Rewrite unit tests for admission code
+7. Modify cluster setup scripts to enable overcommit mode
+8. Testing, debugging race conditions
+9. Prepare and show some QoS demos on the 31st of July community hangout
+--- Extensions ---
+10. Ensure that QoS works for systemd machines
+11. Solve the “restart” issue described above
+

```

Comment:
@bgrant0607 do you know where I can find pointers to how lmctfy resolved the restart issue? The Borg paper mentioned that containers are sent back to the scheduler queue if they fail, but I think we'd want a different (simpler) approach in Kubernetes since we use controllers instead of a scheduler queue.